### DISCUSSION TITLE
trouble with Systemd unit

### DISCUSSION THREAD

**Discussion by sergrosu** [cross-seed/cross-seed#384] [Q&A] [answered]
I installed cross-seed using npm on the following system:
Proxmox container running Ubuntu 20.04. Node version: 16.16.0.

"cross-seed search" works perfectly.

"sudo systemctl start cross-seed" followed by "sudo journalctl -u cross-seed" throws the following error:

Jul 24 15:47:59 Cross-Seed systemd[1]: Started cross-seed daemon.
Jul 24 15:47:59 Cross-Seed node[408]: node:internal/modules/cjs/loader:936
Jul 24 15:47:59 Cross-Seed node[408]:   throw err;
Jul 24 15:47:59 Cross-Seed node[408]:   ^
Jul 24 15:47:59 Cross-Seed node[408]: Error: Cannot find module '/cross-seed'
Jul 24 15:47:59 Cross-Seed node[408]:     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
Jul 24 15:47:59 Cross-Seed node[408]:     at Function.Module._load (node:internal/modules/cjs/loader:778:27)
Jul 24 15:47:59 Cross-Seed node[408]:     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
Jul 24 15:47:59 Cross-Seed node[408]:     at node:internal/main/run_main_module:17:47 {
Jul 24 15:47:59 Cross-Seed node[408]:   code: 'MODULE_NOT_FOUND',
Jul 24 15:47:59 Cross-Seed node[408]:   requireStack: []
Jul 24 15:47:59 Cross-Seed node[408]: }
Jul 24 15:47:59 Cross-Seed systemd[1]: cross-seed.service: Main process exited, code=exited, status=1/FAILURE
Jul 24 15:47:59 Cross-Seed systemd[1]: cross-seed.service: Failed with result 'exit-code'.
Jul 24 15:48:00 Cross-Seed systemd[1]: cross-seed.service: Scheduled restart job, restart counter is at 2.
Jul 24 15:48:00 Cross-Seed systemd[1]: Stopped cross-seed daemon.

It restarts and throws that error again.

I tried it on Ubuntu 21.04 and different node versions as well. The issue remains.

I then tried "ExecStart=/usr/bin/node cross-seed daemon" from the service file directly in terminal and it works.

How do I get the daemon to work? 

Please find attached the config files and log

[cross-seed.zip](https://github.com/mmgoodnow/cross-seed/files/9176223/cross-seed.zip)

**Comment by mmgoodnow** [comment]
Systemd units run in a different context/environment than your shell, with a different PATH. You need to provide an absolute path to `cross-seed`

**Comment by sergrosu** [comment]
I don't see anything regarding that in the installation guide. What's the best way to do it, given a typical installation following your guide?

**Comment by bakerboy448** [comment]
```
# /etc/systemd/system/xseed.service
[Unit]
Description=Cross Seed Search Daemon
After=network.target qbittorrent.service prowlarr.ser>Wants=network-online.target qbittorrent.service prowl>[Service]
User=bakerboy448
Group=bakerboy448
Restart=always
Type=simple
ExecStart=cross-seed daemon
[Install]
WantedBy=multi-user.target
```

Works fine to start the daemon

**Comment by sergrosu** [comment]
@bakerboy448, your version doesn't work for me either. And following @mmgoodnow guidance, you don't have an absolute path to cross-seed either, or am I not seeing something there?

Here's my code for reference:
[Unit]
Description=cross-seed daemon
[Service]
User=khachapu
Group=khachapu
Restart=always
Type=simple
ExecStart=/usr/bin/node cross-seed daemon
[Install]
WantedBy=multi-user.target

**Comment by mmgoodnow** [comment]
The command
```shell
which cross-seed
```

Should help you find an absolute path.

**Comment by sergrosu** [comment]
@mmgoodnow Thanks. The result is /usr/bin/cross-seed

So should the following line work?
ExecStart=/usr/bin/node /usr/bin/cross-seed daemon

**Comment by mmgoodnow** [comment]
Only one way to find out!

**Comment by sergrosu** [comment]
@mmgoodnow, that didn't work either.
New init file:
[Unit]
Description=Cross Seed Search Daemon
[Service]
User=khachapu
Group=khachapu
Restart=always
Type=simple
ExecStart=/usr/bin/node /usr/bin/cross-seed daemon
[Install]
WantedBy=multi-user.target

Same issue:
node:internal/modules/cjs/loader:936
   throw err;
   ^
 Error: Cannot find module '/cross-seed'

So, it looks like even if I provide the absolute path of /usr/bin/cross-seed, it still looks for /cross-seed. Any other ideas why that may be?

**Comment by mmgoodnow** [comment]
Did you install cross-seed globally with `npm install -g`?

**Comment by sergrosu** [comment]
Yes. 

npm list --location=global cross-seed
/usr/lib
`-- cross-seed@4.2.0

**Comment by mmgoodnow** [comment]
I was able to reproduce your error message in my own shell, by doing: 

1. `sudo -s` (opens a root shell)
2. `cd /` (go to the root directory)
3. `/usr/local/bin/node cross-seed`

That got me the `Error: Cannot find module '/cross-seed'` error. 

Given that, I think the systemd unit is still trying to invoke `ExecStart=/usr/bin/node cross-seed daemon`. Did you run `sudo systemctl daemon-reload` after you changed it?

Try changing it to `ExecStart=/usr/bin/node /usr/bin/cross-seed daemon` again (or just verify that it's there) and then run `sudo systemctl daemon-reload`.

**Comment by mmgoodnow** [comment]
You might even be able to shorten it to `ExecStart=/usr/bin/cross-seed daemon` but only if systemd can find where node is. Otherwise, the other way is perfectly fine.

**Comment by sergrosu** [comment]
I was using daemon-reload between the steps. So the problem is definitely not that.

I tried those solutions once again (even if I was positive I tried them before), and it didn't work either... No matter how I change ExecStart paths, the issue is still the same: Error: Cannot find module '/cross-seed'. It looks like it ignores completely the path I provide, which is weird, to say the least.

**Comment by mmgoodnow** [comment, marked as answer]
this really feels like a daemon-reload issue or something similar. Maybe you have multiple unit files, and you're editing the first one and running the second?

I was able to reproduce the error message you're getting, **_only_** with: 

- `ExecStart=/usr/bin/node cross-seed daemon`
- `ExecStart=node cross-seed daemon`

Several other ExecStarts worked perfectly:

- `ExecStart=cross-seed daemon`
- `ExecStart=/usr/bin/cross-seed daemon`
- `ExecStart=node /usr/bin/cross-seed daemon`
- `ExecStart=/usr/bin/node /usr/bin/cross-seed daemon`

**Reply by SuissoNaBisso** [reply]
Hey ! thanks, using one these ExecStart seemed to solve "cannot find module" issue, though I get another error in the journal saying i'm missint torrentDir while it's actually well set in my config file, and when running "cross-seed search" everyhting works fine. So I guess the daemon is not launching using config file, any idea on how can I solve this ? Thanks
